From d597cf2c1a9940db1457e052dc5b119a988fa136 Mon Sep 17 00:00:00 2001
From: Tim Carney <tim.carney@cargt.com>
Date: Wed, 22 Apr 2026 13:21:56 -0500
Subject: [PATCH] Add support for the GLT028240320IS1 display on the Cargt
 00365 carrier with the i.MX 93 00363 OSM-L SOM. This includes device tree and
 DRM driver support for Sitronix ST7789T3 display panels.

---
 .../bindings/display/sitronix,st7789t3.yaml   |  72 ++++
 arch/arm64/boot/dts/freescale/Makefile        |   3 +-
 ...mx93-cargt-00363-00365-glt028240320is1.dts |  13 +
 ...x93-cargt-00363-00365-glt028240320is1.dtsi |  89 +++++
 drivers/gpu/drm/tiny/Kconfig                  |  14 +
 drivers/gpu/drm/tiny/Makefile                 |   1 +
 drivers/gpu/drm/tiny/st7789t3.c               | 328 ++++++++++++++++++
 7 files changed, 519 insertions(+), 1 deletion(-)
 create mode 100644 Documentation/devicetree/bindings/display/sitronix,st7789t3.yaml
 create mode 100644 arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dts
 create mode 100644 arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dtsi
 create mode 100644 drivers/gpu/drm/tiny/st7789t3.c

diff --git a/Documentation/devicetree/bindings/display/sitronix,st7789t3.yaml b/Documentation/devicetree/bindings/display/sitronix,st7789t3.yaml
new file mode 100644
index 000000000000..54378cf3d94a
--- /dev/null
+++ b/Documentation/devicetree/bindings/display/sitronix,st7789t3.yaml
@@ -0,0 +1,72 @@
+# SPDX-License-Identifier: GPL-2.0-only
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/display/sitronix,st7735r.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Sitronix ST7789T3 Display Panels
+
+maintainers:
+  - Tim Carney <tim.carney@cargt.com>
+
+description:
+  This binding is for display panels using a Sitronix ST7789T3
+  controller in SPI mode. It is based on the Sitronix ST7735R
+  Display Panel driver from David Lechner <david@lechnology.com>
+
+
+allOf:
+  - $ref: panel/panel-common.yaml#
+  - $ref: /schemas/spi/spi-peripheral-props.yaml#
+
+properties:
+  compatible:
+    oneOf:
+      - description:
+          GT Display glt028240320IS1 2.8" 320x240 Color TFT LCD
+        items:
+          - enum:
+              - gtdisplay,glt028240320IS1
+          - const: sitronix,st7789t3
+
+  dc-gpios:
+    maxItems: 1
+    description: Display data/command selection (D/CX)
+
+  backlight: true
+  reg: true
+  spi-max-frequency: true
+  reset-gpios: true
+  rotation: true
+  disp_inversion: true
+
+required:
+  - compatible
+  - reg
+  - dc-gpios
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    backlight: backlight {
+            compatible = "gpio-backlight";
+            gpios = <&gpio 44 GPIO_ACTIVE_HIGH>;
+    };
+
+    spi {
+            #address-cells = <1>;
+            #size-cells = <0>;
+
+            display@0{
+              compatible = "gtdisplay,glt028240320IS1", "sitronix,st7789t3";
+              reg = <0>;
+              spi-max-frequency = <24000000>;
+              dc-gpios = <&gpio2 4 GPIO_ACTIVE_HIGH>;
+              reset-gpios = <&pca9555 1 GPIO_ACTIVE_HIGH>;
+            };
+    };
+
+...
diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index f7077bfffd3f..d56edd4660ab 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -446,10 +446,11 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-tqma9352-mba93xxla.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx93-cargt-00324-00326.dtb \
 				imx93-cargt-00324-00326-glt0701024600is2.dtb \
 				imx93-cargt-00324-00326-glt1011280800is1.dtb \
 				imx93-cargt-00359-00406.dtb \
 				imx93-cargt-00363-00365.dtb \
 				imx93-cargt-00363-00365-glt0557201280is1.dtb \
-				imx93-cargt-00363-00365-glt1011280800is1.dtb
+				imx93-cargt-00363-00365-glt1011280800is1.dtb \
+				imx93-cargt-00363-00365-glt028240320is1.dtb

 dtb-$(CONFIG_ARCH_MXC) += imx91-9x9-qsb.dtb \
 			  imx91-9x9-qsb-can1.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dts b/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dts
new file mode 100644
index 000000000000..cfa35a53cae6
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dts
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2022 NXP
+ */
+
+/dts-v1/;
+
+#include "imx93-cargt-00363-00365.dts"
+#include "imx93-cargt-00363-00365-glt028240320is1.dtsi"
+
+/ {
+	model = "Cargt i.MX93 00363 OSM-L SOM on 00365 carrier board with 2.8in SPI Display(GLT028240320IS1-CTP)";
+};
diff --git a/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dtsi b/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dtsi
new file mode 100644
index 000000000000..0fc49a49fe49
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx93-cargt-00363-00365-glt028240320is1.dtsi
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2022 NXP
+ */
+
+/ {
+	pwm_backlight: pwm_backlight {
+		compatible = "pwm-backlight";
+		pwms = <&tpm6 1 5000000 0>; // 200 Hz
+		brightness-levels = <0 100>;
+		num-interpolated-steps = <100>;
+		default-brightness-level = <80>;
+		status = "okay";
+	};
+};
+
+&lpspi8 {
+	fsl,spi-num-chipselects = <1>;
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&pinctrl_lpspi8>;
+	pinctrl-1 = <&pinctrl_lpspi8>;
+	cs-gpios = <&gpio2 12 GPIO_ACTIVE_LOW>;
+	status = "okay";
+
+	/delete-property/ dmas;
+	/delete-property/ dma-names;
+
+	display@0{
+		compatible = "gtdisplay,glt028240320IS1", "sitronix,st7789t3";
+		reg = <0>;
+		spi-max-frequency = <12000000>;
+		dc-gpios = <&gpio1 7 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&pca9555_20 8 GPIO_ACTIVE_HIGH>;
+		backlight = <&pwm_backlight>;
+	};
+};
+
+
+&lpuart2 {
+	status = "disabled";
+};
+
+&tpm6 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_backlight_pwm>;
+	status = "okay";
+};
+
+&lpi2c2 {
+	focaltech_touch: touchscreen@38 {
+		compatible = "focaltech,ft6236";
+		reg = <0x38>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_touch_io>;
+		interrupt-parent = <&gpio2>;
+		interrupts = <8 IRQ_TYPE_EDGE_FALLING>;
+        touchscreen-size-x = <240>;
+        touchscreen-size-y = <320>;
+		touchscreen-inverted-y;
+		reset-gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+		wakeup-source;
+      };
+};
+
+&iomuxc {
+	pinctrl_lpspi8: lpspi8grp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO12__GPIO2_IO12      0x3fe
+			MX93_PAD_GPIO_IO13__LPSPI8_SIN      0x3fe
+			MX93_PAD_GPIO_IO14__LPSPI8_SOUT     0x3fe
+			MX93_PAD_GPIO_IO15__LPSPI8_SCK      0x3fe
+            MX93_PAD_UART2_TXD__GPIO1_IO07      0x3fe // Data Command Switch
+		>;
+	};
+
+	pinctrl_backlight_pwm: backlight-pwmgrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO23__TPM6_CH1		0x2
+		>;
+	};
+
+	pinctrl_touch_io: touchiogrp {
+		fsl,pins = <
+			MX93_PAD_UART2_RXD__GPIO1_IO06		0x3fe // RST
+			MX93_PAD_GPIO_IO08__GPIO2_IO08		0x3fe // INT
+		>;
+	};
+};
\ No newline at end of file
diff --git a/drivers/gpu/drm/tiny/Kconfig b/drivers/gpu/drm/tiny/Kconfig
index f6889f649bc1..fbfbd6151f77 100644
--- a/drivers/gpu/drm/tiny/Kconfig
+++ b/drivers/gpu/drm/tiny/Kconfig
@@ -212,3 +212,17 @@ config TINYDRM_ST7735R
 	  * Okaya RH128128T 1.44" 128x128 TFT

 	  If M is selected the module will be called st7735r.
+
+config TINYDRM_ST7789T3
+	tristate "DRM support for Sitronix ST7789T3 display panels"
+	depends on DRM && SPI
+	select DRM_KMS_HELPER
+	select DRM_GEM_DMA_HELPER
+	select DRM_MIPI_DBI
+	select BACKLIGHT_CLASS_DEVICE
+	help
+	  DRM driver for Sitronix ST7789T3 with one of the following
+	  LCDs:
+	  * GT Display glt028240320IS1 2.8" 320x240 TFT
+
+	  If M is selected the module will be called st7789t3.
diff --git a/drivers/gpu/drm/tiny/Makefile b/drivers/gpu/drm/tiny/Makefile
index 76dde89a044b..826dea30620e 100644
--- a/drivers/gpu/drm/tiny/Makefile
+++ b/drivers/gpu/drm/tiny/Makefile
@@ -16,3 +16,4 @@ obj-$(CONFIG_TINYDRM_MI0283QT)		+= mi0283qt.o
 obj-$(CONFIG_TINYDRM_REPAPER)		+= repaper.o
 obj-$(CONFIG_TINYDRM_ST7586)		+= st7586.o
 obj-$(CONFIG_TINYDRM_ST7735R)		+= st7735r.o
+obj-$(CONFIG_TINYDRM_ST7789T3)		+= st7789t3.o
diff --git a/drivers/gpu/drm/tiny/st7789t3.c b/drivers/gpu/drm/tiny/st7789t3.c
new file mode 100644
index 000000000000..b2728fc49757
--- /dev/null
+++ b/drivers/gpu/drm/tiny/st7789t3.c
@@ -0,0 +1,328 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * DRM driver for display panels connected to a Sitronix ST7789T3
+ * display controller in SPI mode.
+ * https://files.waveshare.com/wiki/2.8inch-Capacitive-Touch-LCD/ST7789T3_SPEC_V1.0.pdf
+ *
+ * Based on the DRM driver for display panels connected to a Sitronix ST7715R or ST7735R
+ * Copyright 2017 David Lechner <david@lechnology.com>
+ * Copyright (C) 2019 Glider bvba
+ */
+
+#include <linux/backlight.h>
+#include <linux/delay.h>
+#include <linux/dma-buf.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/property.h>
+#include <linux/spi/spi.h>
+#include <video/mipi_display.h>
+
+#include <drm/drm_atomic_helper.h>
+#include <drm/drm_drv.h>
+#include <drm/drm_fbdev_generic.h>
+#include <drm/drm_gem_atomic_helper.h>
+#include <drm/drm_gem_dma_helper.h>
+#include <drm/drm_managed.h>
+#include <drm/drm_mipi_dbi.h>
+
+#define ST7789_CMD_NOP             0x00        /**< no operation command */
+#define ST7789_CMD_SWRESET         0x01        /**< software reset command */
+#define ST7789_CMD_SLPIN           0x10        /**< sleep in command */
+#define ST7789_CMD_SLPOUT          0x11        /**< sleep out command */
+#define ST7789_CMD_PTLON           0x12        /**< partial mode on command */
+#define ST7789_CMD_NORON           0x13        /**< normal display mode on command */
+#define ST7789_CMD_INVOFF          0x20        /**< display inversion off command */
+#define ST7789_CMD_INVON           0x21        /**< display inversion on command */
+#define ST7789_CMD_GAMSET          0x26        /**< display inversion set command */
+#define ST7789_CMD_DISPOFF         0x28        /**< display off command */
+#define ST7789_CMD_DISPON          0x29        /**< display on command */
+#define ST7789_CMD_CASET           0x2A        /**< column address set command */
+#define ST7789_CMD_RASET           0x2B        /**< row address set command */
+#define ST7789_CMD_RAMWR           0x2C        /**< memory write command */
+#define ST7789_CMD_PTLAR           0x30        /**< partial start/end address set command */
+#define ST7789_CMD_VSCRDEF         0x33        /**< vertical scrolling definition command */
+#define ST7789_CMD_TEOFF           0x34        /**< tearing effect line off command */
+#define ST7789_CMD_TEON            0x35        /**< tearing effect line on command */
+#define ST7789_CMD_MADCTL          0x36        /**< memory data access control command */
+#define ST7789_CMD_VSCRSADD        0x37        /**< vertical scrolling start address command */
+#define ST7789_CMD_IDMOFF          0x38        /**< idle mode off command */
+#define ST7789_CMD_IDMON           0x39        /**< idle mode on command */
+#define ST7789_CMD_COLMOD          0x3A        /**< interface pixel format command */
+#define ST7789_CMD_RAMWRC          0x3C        /**< memory write continue command */
+#define ST7789_CMD_TESCAN          0x44        /**< set tear scanline command */
+#define ST7789_CMD_WRDISBV         0x51        /**< write display brightness command */
+#define ST7789_CMD_WRCTRLD         0x53        /**< write CTRL display command */
+#define ST7789_CMD_WRCACE          0x55        /**< write content adaptive brightness control and color enhancement command */
+#define ST7789_CMD_WRCABCMB        0x5E        /**< write CABC minimum brightness command */
+#define ST7789_CMD_RAMCTRL         0xB0        /**< ram control command */
+#define ST7789_CMD_RGBCTRL         0xB1        /**< rgb control command */
+#define ST7789_CMD_PORCTRL         0xB2        /**< porch control command */
+#define ST7789_CMD_FRCTRL1         0xB3        /**< frame rate control 1 command */
+#define ST7789_CMD_PARCTRL         0xB5        /**< partial mode control command */
+#define ST7789_CMD_GCTRL           0xB7        /**< gate control command */
+#define ST7789_CMD_GTADJ           0xB8        /**< gate on timing adjustment command */
+#define ST7789_CMD_DGMEN           0xBA        /**< digital gamma enable command */
+#define ST7789_CMD_VCOMS           0xBB        /**< vcoms setting command */
+#define ST7789_CMD_LCMCTRL         0xC0        /**< lcm control command */
+#define ST7789_CMD_IDSET           0xC1        /**< id setting command */
+#define ST7789_CMD_VDVVRHEN        0xC2        /**< vdv and vrh command enable command */
+#define ST7789_CMD_VRHS            0xC3        /**< vrh set command */
+#define ST7789_CMD_VDVSET          0xC4        /**< vdv setting command */
+#define ST7789_CMD_VCMOFSET        0xC5        /**< vcoms offset set command */
+#define ST7789_CMD_FRCTR2          0xC6        /**< fr control 2 command */
+#define ST7789_CMD_CABCCTRL        0xC7        /**< cabc control command */
+#define ST7789_CMD_REGSEL1         0xC8        /**< register value selection1 command */
+#define ST7789_CMD_REGSEL2         0xCA        /**< register value selection2 command */
+#define ST7789_CMD_PWMFRSEL        0xCC        /**< pwm frequency selection command */
+#define ST7789_CMD_PWCTRL1         0xD0        /**< power control 1 command */
+#define ST7789_CMD_VAPVANEN        0xD2        /**< enable vap/van signal output command */
+#define ST7789_CMD_CMD2EN          0xDF        /**< command 2 enable command */
+#define ST7789_CMD_PVGAMCTRL       0xE0        /**< positive voltage gamma control command */
+#define ST7789_CMD_NVGAMCTRL       0xE1        /**< negative voltage gamma control command */
+#define ST7789_CMD_DGMLUTR         0xE2        /**< digital gamma look-up table for red command */
+#define ST7789_CMD_DGMLUTB         0xE3        /**< digital gamma look-up table for blue command */
+#define ST7789_CMD_GATECTRL        0xE4        /**< gate control command */
+#define ST7789_CMD_SPI2EN          0xE7        /**< spi2 command */
+#define ST7789_CMD_PWCTRL2         0xE8        /**< power control 2 command */
+#define ST7789_CMD_EQCTRL          0xE9        /**< equalize time control command */
+#define ST7789_CMD_PROMCTRL        0xEC        /**< program control command */
+#define ST7789_CMD_PROMEN          0xFA        /**< program mode enable command */
+#define ST7789_CMD_NVMSET          0xFC        /**< nvm setting command */
+#define ST7789_CMD_PROMACT         0xFE        /**< program action command */
+
+#define ST7789T3_MY	BIT(7)
+#define ST7789T3_MX	BIT(6)
+#define ST7789T3_MV	BIT(5)
+#define ST7789T3_RGB	BIT(3)
+
+struct st7789t3_cfg {
+	const struct drm_display_mode mode;
+	unsigned int left_offset;
+	unsigned int top_offset;
+	unsigned int write_only:1;
+	unsigned int rgb:1;		/* RGB (vs. BGR) */
+	unsigned int disp_inversion:1;
+};
+
+struct st7789t3_priv {
+	struct mipi_dbi_dev dbidev;	/* Must be first for .release() */
+	const struct st7789t3_cfg *cfg;
+};
+
+static void st7789t3_pipe_enable(struct drm_simple_display_pipe *pipe,
+				struct drm_crtc_state *crtc_state,
+				struct drm_plane_state *plane_state)
+{
+	struct mipi_dbi_dev *dbidev = drm_to_mipi_dbi_dev(pipe->crtc.dev);
+	struct st7789t3_priv *priv = container_of(dbidev, struct st7789t3_priv,
+						 dbidev);
+	struct mipi_dbi *dbi = &dbidev->dbi;
+	int ret, idx;
+	u8 addr_mode;
+
+	if (!drm_dev_enter(pipe->crtc.dev, &idx))
+		return;
+
+	DRM_DEBUG_KMS("\n");
+
+	ret = mipi_dbi_poweron_reset(dbidev);
+	if (ret)
+		goto out_exit;
+
+	msleep(150);
+
+	mipi_dbi_command(dbi, MIPI_DCS_EXIT_SLEEP_MODE);
+	msleep(500);
+
+	mipi_dbi_command(dbi, ST7789_CMD_RGBCTRL, 0x01, 0x2c, 0x2d);
+	mipi_dbi_command(dbi, ST7789_CMD_PORCTRL, 0x01, 0x2c, 0x2d);
+	mipi_dbi_command(dbi, ST7789_CMD_FRCTRL1, 0x01, 0x2c, 0x2d, 0x01, 0x2c, 0x2d);
+	mipi_dbi_command(dbi, ST7789_CMD_LCMCTRL, 0xa2, 0x02, 0x84);
+	mipi_dbi_command(dbi, ST7789_CMD_IDSET, 0xc5);
+	mipi_dbi_command(dbi, ST7789_CMD_VDVVRHEN, 0x01);   //9.2.12 VDVVRHEN (C2h): VDV and VRH Command Enable
+	mipi_dbi_command(dbi, ST7789_CMD_VRHS, 0x10);   //9.2.13 VRHS (C3h): VRH Set
+	mipi_dbi_command(dbi, ST7789_CMD_VRHS, 0x10);   //9.2.13 VRHS (C3h): VRH Sett
+	mipi_dbi_command(dbi, ST7789_CMD_FRCTR2, 0x0F);   //9.2.16 FRCTRL2 (C6h): Frame Rate Control in Normal Mode
+	mipi_dbi_command(dbi, ST7789_CMD_PWCTRL1, 0xA4, 0xA1);   //9.2.21 PWCTRL1 (D0h): Power Control 1
+	mipi_dbi_command(dbi, ST7789_CMD_VCMOFSET, 0x0e); //0xc5
+	mipi_dbi_command(dbi, MIPI_DCS_EXIT_INVERT_MODE); //ST7789_CMD_INVOFF
+
+	switch (dbidev->rotation) {
+	default:
+		addr_mode = ST7789T3_MX | ST7789T3_MY;
+		break;
+	case 90:
+		addr_mode = ST7789T3_MX | ST7789T3_MV;
+		break;
+	case 180:
+		addr_mode = 0;
+		break;
+	case 270:
+		addr_mode = ST7789T3_MY | ST7789T3_MV;
+		break;
+	}
+
+	if (priv->cfg->rgb)
+		addr_mode |= ST7789T3_RGB;
+
+	mipi_dbi_command(dbi, MIPI_DCS_SET_ADDRESS_MODE, addr_mode); //ST7789_CMD_MADCTL
+	mipi_dbi_command(dbi, MIPI_DCS_SET_PIXEL_FORMAT, MIPI_DCS_PIXEL_FMT_16BIT); //ST7789_CMD_COLMOD
+	mipi_dbi_command(dbi, ST7789_CMD_PVGAMCTRL, 0xD0, 0x07, 0x0E, 0x0B, 0x0A, 0x14, 0x38, 0x33, 0x4F, 0x37, 0x16, 0x16, 0x2A, 0x2E); //9.2.24 PVGAMCTRL (E0h): Positive Voltage Gamma Control
+	mipi_dbi_command(dbi, ST7789_CMD_NVGAMCTRL, 0xD0, 0x0B, 0x10, 0x08, 0x08, 0x06, 0x35, 0x54, 0x4D, 0x0A, 0x14, 0x14, 0x2C, 0x2F); //9.2.25 NVGAMCTRL (E1h): Negative Voltage Gamma Control
+	mipi_dbi_command(dbi, ST7789_CMD_EQCTRL, 0x11, 0x11, 0x03); //9.2.31 EQCTRL (E9h): Equalize time control
+
+	if (priv->cfg->disp_inversion)
+	{
+ 		mipi_dbi_command(dbi, ST7789_CMD_INVON);
+	}
+
+	mipi_dbi_command(dbi, MIPI_DCS_SET_DISPLAY_ON); //ST7789_CMD_DISPON
+
+	msleep(100);
+
+	mipi_dbi_command(dbi, MIPI_DCS_ENTER_NORMAL_MODE); //ST7789_CMD_NORON
+
+	msleep(20);
+
+	mipi_dbi_enable_flush(dbidev, crtc_state, plane_state);
+out_exit:
+	drm_dev_exit(idx);
+}
+
+static const struct drm_simple_display_pipe_funcs st7789t3_pipe_funcs = {
+	DRM_MIPI_DBI_SIMPLE_DISPLAY_PIPE_FUNCS(st7789t3_pipe_enable),
+};
+
+static const struct st7789t3_cfg gtg_glt028240320IS1_cfg = {
+	.mode			= { DRM_SIMPLE_MODE(320, 240, 67, 35) },
+	.left_offset	= 0,
+	.top_offset		= 0,
+	.rgb			= false,
+	.write_only     = true,
+	.disp_inversion = true,
+};
+
+DEFINE_DRM_GEM_DMA_FOPS(st7789t3_fops);
+
+static const struct drm_driver st7789t3_driver = {
+	.driver_features	= DRIVER_GEM | DRIVER_MODESET | DRIVER_ATOMIC,
+	.fops			= &st7789t3_fops,
+	DRM_GEM_DMA_DRIVER_OPS_VMAP,
+	.debugfs_init		= mipi_dbi_debugfs_init,
+	.name			= "st7789t3",
+	.desc			= "Sitronix ST7789T3",
+	.date			= "20171128",
+	.major			= 1,
+	.minor			= 0,
+};
+
+static const struct of_device_id st7789t3_of_match[] = {
+	{ .compatible = "gtdisplay,glt028240320IS1", .data = &gtg_glt028240320IS1_cfg },
+	{ },
+};
+MODULE_DEVICE_TABLE(of, st7789t3_of_match);
+
+static const struct spi_device_id st7789t3_id[] = {
+	{ "glt028240320IS1", (uintptr_t)&gtg_glt028240320IS1_cfg },
+	{ },
+};
+MODULE_DEVICE_TABLE(spi, st7789t3_id);
+
+static int st7789t3_probe(struct spi_device *spi)
+{
+	struct device *dev = &spi->dev;
+	const struct st7789t3_cfg *cfg;
+	struct mipi_dbi_dev *dbidev;
+	struct st7789t3_priv *priv;
+	struct drm_device *drm;
+	struct mipi_dbi *dbi;
+	struct gpio_desc *dc;
+	u32 rotation = 0;
+	int ret;
+
+	cfg = device_get_match_data(&spi->dev);
+	if (!cfg)
+		cfg = (void *)spi_get_device_id(spi)->driver_data;
+
+	priv = devm_drm_dev_alloc(dev, &st7789t3_driver,
+				  struct st7789t3_priv, dbidev.drm);
+	if (IS_ERR(priv))
+		return PTR_ERR(priv);
+
+	dbidev = &priv->dbidev;
+	priv->cfg = cfg;
+
+	dbi = &dbidev->dbi;
+	drm = &dbidev->drm;
+
+	dbi->reset = devm_gpiod_get(dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(dbi->reset))
+		return dev_err_probe(dev, PTR_ERR(dbi->reset), "Failed to get GPIO 'reset'\n");
+
+	dc = devm_gpiod_get(dev, "dc", GPIOD_OUT_LOW);
+	if (IS_ERR(dc))
+		return dev_err_probe(dev, PTR_ERR(dc), "Failed to get GPIO 'dc'\n");
+
+	dbidev->backlight = devm_of_find_backlight(dev);
+	if (IS_ERR(dbidev->backlight))
+		return PTR_ERR(dbidev->backlight);
+
+	device_property_read_u32(dev, "rotation", &rotation);
+
+	ret = mipi_dbi_spi_init(spi, dbi, dc);
+	if (ret)
+		return ret;
+
+	if (cfg->write_only)
+		dbi->read_commands = NULL;
+
+	dbidev->left_offset = cfg->left_offset;
+	dbidev->top_offset = cfg->top_offset;
+
+	ret = mipi_dbi_dev_init(dbidev, &st7789t3_pipe_funcs, &cfg->mode,
+				rotation);
+	if (ret)
+		return ret;
+
+	drm_mode_config_reset(drm);
+
+	ret = drm_dev_register(drm, 0);
+	if (ret)
+		return ret;
+
+	spi_set_drvdata(spi, drm);
+
+	drm_fbdev_generic_setup(drm, 0);
+
+	return 0;
+}
+
+static void st7789t3_remove(struct spi_device *spi)
+{
+	struct drm_device *drm = spi_get_drvdata(spi);
+
+	drm_dev_unplug(drm);
+	drm_atomic_helper_shutdown(drm);
+}
+
+static void st7789t3_shutdown(struct spi_device *spi)
+{
+	drm_atomic_helper_shutdown(spi_get_drvdata(spi));
+}
+
+static struct spi_driver st7789t3_spi_driver = {
+	.driver = {
+		.name = "st7789t3",
+		.of_match_table = st7789t3_of_match,
+	},
+	.id_table = st7789t3_id,
+	.probe = st7789t3_probe,
+	.remove = st7789t3_remove,
+	.shutdown = st7789t3_shutdown,
+};
+module_spi_driver(st7789t3_spi_driver);
+
+MODULE_DESCRIPTION("Sitronix ST7789T3 DRM driver");
+MODULE_AUTHOR("Tim Carney <tim.carney@cargt.com>");
+MODULE_LICENSE("GPL");
